src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,551 @@
+                _randomByteGenerator = Ensure.IsNotNull(randomByteGenerator, nameof(randomByteGenerator));
+                _dateTimeProvider = Ensure.IsNotNull(dateTimeProvider, nameof(dateTimeProvider));
+            }
+
+            public string Name
+            {
+                get { return MechanismName; }
+            }
+
+            public ISaslStep Initialize(IConnection connection, SaslConversation conversation, ConnectionDescription description)
+            {
+                Ensure.IsNotNull(connection, nameof(connection));
+                Ensure.IsNotNull(description, nameof(description));
+
+                var nonce = GenerateRandomBytes();
+
+                var document = new BsonDocument()
+                    .Add("r", new BsonBinaryData(nonce))
+                    .Add("p", new BsonInt32('n'));
+
+                var clientFirstMessageBytes = ToBytes(document);
+
+                return new ClientFirst(clientFirstMessageBytes, nonce, _credential, _sessionToken, _dateTimeProvider);
+            }
+
+            private byte[] GenerateRandomBytes()
+            {
+                return _randomByteGenerator.Generate(_clientNonceLength);
+            }
+        }
+
+        private class ClientFirst : ISaslStep
+        {
+            private readonly byte[] _bytesToSendToServer;
+            private readonly UsernamePasswordCredential _credential;
+            private readonly IClock _dateTimeProvider;
+            private readonly byte[] _nonce;
+            private readonly string _sessionToken;
+
+            public ClientFirst(
+                byte[] bytesToSendToServer,
+                byte[] nonce,
+                UsernamePasswordCredential credential,
+                string sessionToken,
+                IClock dateTimeProvider)
+            {
+                _bytesToSendToServer = bytesToSendToServer;
+                _credential = credential;
+                _dateTimeProvider = dateTimeProvider;
+                _nonce = nonce;
+                _sessionToken = sessionToken;
+            }
+
+            public byte[] BytesToSendToServer
+            {
+                get { return _bytesToSendToServer; }
+            }
+
+            public bool IsComplete
+            {
+                get { return false; }
+            }
+
+            public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedFromServer)
+            {
+                var serverFirstMessageDoc = ToDocument(bytesReceivedFromServer);
+                var serverNonce = serverFirstMessageDoc["s"].AsByteArray;
+                var host = serverFirstMessageDoc["h"].AsString;
+
+                if (serverNonce.Length != _clientNonceLength * 2 || !serverNonce.Take(_clientNonceLength).SequenceEqual(_nonce))
+                {
+                    throw new MongoAuthenticationException(conversation.ConnectionId, message: "Server sent an invalid nonce.");
+                }
+
+                AwsSignatureVersion4.SignRequest(
```

Comment:
I'm not sure that `SignRequest` method name really tells about what happens inside. As I understand, we just **create** an authenticateHeader